kp_registry/routers/kps.py

```diff
@@ -48,39 +45,148 @@ async def get_knowledge_provider(
+                for endpoint in endpoints
+            ],
+            return_exceptions=True,
+        )
+    meta_kgs = []
+    for endpoint, response in zip(endpoints, responses):
+        if isinstance(response, Exception):
+            LOGGER.warning(
+                "Error accessing /meta_knowledge_graph for %s (https://smart-api.info/registry?q=%s): %s",
+                endpoint["title"],
+                _id,
+                response,
+            )
+            continue
+        if response.status_code >= 300:
+            LOGGER.warning(
+                "Bad response from /meta_knowledge_graph for %s (https://smart-api.info/registry?q=%s): %s",
+                endpoint["title"],
+                _id,
+                f"<HTTP {response.status_code}> {response.text}",
+            )
+            continue
+        try:
+            # validate /meta_knowledge_graph response.
+            MetaKnowledgeGraph(**response.json())
+        except json.decoder.JSONDecodeError as err:
+            LOGGER.warning(
+                "Error decoding /meta_knowledge_graph response for %s (https://smart-api.info/registry?q=%s): %s",
+                endpoint["title"],
+                _id,
+                response.text,
+            )
+            continue
+        except pydantic.ValidationError as err:
+            LOGGER.warning(
+                "Failed to validate /meta_knowledge_graph response for %s (https://smart-api.info/registry?q=%s): %s",
+                endpoint["title"],
+                _id,
+                err,
+            )
+            continue
+        meta_kgs.append((endpoint, response.json()))
+
+    kps = dict()
+    for endpoint, meta_kg in meta_kgs:
+        try:
+            kps[endpoint["title"]] = {
+                "url": endpoint["url"],
+                "operations": meta_kg["edges"],
+                "details": {"preferred_prefixes": {
+                    category: value["id_prefixes"]
+                    for category, value in meta_kg["nodes"].items()
+                }},
+            }
+        except Exception as err:
+            tb = traceback.format_exc()
+            LOGGER.warning(
+                "Error parsing KP details for %s (https://smart-api.info/registry?q=%s): %s",
+                endpoint["title"],
+                _id,
+                tb,
+            )
+    async with Registry(settings.db_uri) as registry:
+        await registry.delete_all()
+        await registry.add(**kps)
+    LOGGER.debug("Reloaded registry.")
 
 
-@router.post('/clear', status_code=status.HTTP_204_NO_CONTENT)
-async def clear_kps(
-    registry=Depends(get_registry),
-):
-    """Clear all registered KPs."""
-    await registry.delete_all()
+@router.post('/refresh', status_code=status.HTTP_202_ACCEPTED)
+async def refresh_kps(background_tasks: BackgroundTasks):
+    """Refresh registered KPs by consulting SmartAPI registry."""
+    background_tasks.add_task(load_from_smartapi)
```

Comment:
Should this run on startup? I think it makes sense to - we don't really ever use the KP registry without running this.

Reply:
I had that thought and actually had it implemented, but then decided that the process of hitting all these external services should be very deliberately triggered rather than run on startup.